doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,245 @@
+The `frames` are pointers to the individual frames that make up the packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+struct ossl_quic_packet_st {
+    SSL *connection;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` struture from the ACK Manager overlaps this significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in the Application Data number space?
+- the `packet_header` will become a structure / union most likely.
+
+## Interactions
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+### Stream Send Buffers
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL *stream, const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
+
+#### Notes
+
+- Should more than one frame be queueable at once?
+- More arguments are likely going to be required.
+- How to destroy frames?  `OPENSSL_free` or a callback?
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+#### Notes
+
+- How distinguish handshake from rest?
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/* API to be defined for flow controller */
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/* Pluggable congestion controller APIs go here */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+__owur int ossl_quic_send_packets(SSL *connection);
+```
```

Comment:
I'd like to see the input and output interfaces consumed by the TX packetiser more clearly delineated. e.g. How is the packetiser instantiated? How and what dependencies are injected (e.g. the network BIO). Is this module the exclusive permitted user of BIO_sendmmsg? (Presumably it is.)

Will the TX packetiser need to handle timer events and have the event queue provided to it (e.g. frame/packet coalescing delays)?

Reply:
> Is this module the exclusive permitted user of BIO_sendmmsg? (Presumably it is.)

At least in the original design the TX QUIC record layer would call the BIO_sendmmsg(). However that does not really work well with 1 call to record layer for QUIC 1 packet and it also does not work for packet coalescing. So... this is still unclear.

On the other hand I suppose some clever filtering BIO inserted in front of the real BIO_dgram might work.